pkg/api/controller.go

```diff
@@ -2845,6 +2846,53 @@ func (c *Controller) logCommitsHelper(w http.ResponseWriter, r *http.Request, re
 	writeResponse(w, http.StatusOK, response)
 }
 
+func (c *Controller) HeadObject(w http.ResponseWriter, r *http.Request, repository string, ref string, params HeadObjectParams) {
+	if !c.authorize(w, r, permissions.Node{
+		Permission: permissions.Permission{
+			Action:   permissions.ReadObjectAction,
+			Resource: permissions.ObjectArn(repository, params.Path),
+		},
+	}) {
+		return
+	}
+	ctx := r.Context()
+	c.LogAction(ctx, "head_object", r, repository, ref, "")
+
+	// read the FS entry
+	entry, err := c.Catalog.GetEntry(ctx, repository, ref, params.Path, catalog.GetEntryParams{ReturnExpired: true})
+	rr := httptest.NewRecorder()
+	if c.handleAPIError(ctx, w, err) {
+		w.WriteHeader(rr.Code)
+		return
```

Comment:
Extract the api error handle so you can capture the code and have the same result without using a test library to record the response write generate.
```go
func (c *Controller) handleAPIErrorCallback(ctx context.Context, w http.ResponseWriter, err error, cb func(w http.ResponseWriter, code int, v interface{})) bool {
	switch {
	case errors.Is(err, catalog.ErrNotFound),
		errors.Is(err, graveler.ErrNotFound),
		errors.Is(err, actions.ErrNotFound),
		errors.Is(err, auth.ErrNotFound),
		errors.Is(err, kv.ErrNotFound):
		cb(w, http.StatusNotFound, err)

	case errors.Is(err, graveler.ErrDirtyBranch),
		errors.Is(err, graveler.ErrCommitMetaRangeDirtyBranch),
		errors.Is(err, catalog.ErrNoDifferenceWasFound),
		errors.Is(err, graveler.ErrNoChanges),
		errors.Is(err, permissions.ErrInvalidServiceName),
		errors.Is(err, permissions.ErrInvalidAction),
		errors.Is(err, model.ErrValidationError),
		errors.Is(err, graveler.ErrInvalidRef),
		errors.Is(err, graveler.ErrInvalidValue),
		errors.Is(err, actions.ErrParamConflict):
		cb(w, http.StatusBadRequest, err)

	case errors.Is(err, graveler.ErrNotUnique),
		errors.Is(err, graveler.ErrConflictFound),
		errors.Is(err, graveler.ErrRevertMergeNoParent):
		cb(w, http.StatusConflict, err)

	case errors.Is(err, catalog.ErrFeatureNotSupported):
		cb(w, http.StatusNotImplemented, err)

	case errors.Is(err, graveler.ErrLockNotAcquired):
		cb(w, http.StatusInternalServerError, "branch is currently locked, try again later")

	case errors.Is(err, adapter.ErrDataNotFound):
		cb(w, http.StatusGone, "No data")

	case errors.Is(err, auth.ErrAlreadyExists):
		cb(w, http.StatusConflict, "Already exists")

	case errors.Is(err, graveler.ErrTooManyTries):
		cb(w, http.StatusLocked, "Too many attempts, try again later")

	case err != nil:
		c.Logger.WithContext(ctx).WithError(err).Error("API call returned status internal server error")
		cb(w, http.StatusInternalServerError, err)

	default:
		return false
	}

	return true
}

func (c *Controller) handleAPIError(ctx context.Context, w http.ResponseWriter, err error) bool {
	return c.handleAPIErrorCallback(ctx, w, err, writeError)
}

```

Reply:
Done! did the same for `c.authorize`.